index.bs

```diff
@@ -3406,6 +3408,58 @@ The [=remote end steps=] with |command parameters| are:
 
 </div>
 
+#### The browsingContext.setUserAgent Command ####  {#command-browsingContext-setUserAgent}
+
+The <dfn export for=commands>browsingContext.setUserAgent</dfn> command sets the
+{{NavigatorID/userAgent}} string for a browsing context.
+
+<dl>
+   <dt>Command Type</dt>
+   <dd>
+    <pre class="cddl remote-cddl">
+      browsingContext.SetUserAgent = (
+        method: "browsingContext.setUserAgent",
+        params: browsingContext.SetUserAgentParameters
+      )
+
+      browsingContext.SetUserAgentParameters = {
+        context: browsingContext.BrowsingContext,
+        ? userAgent: text,
+      }
+    </pre>
+   </dd>
+   <dt>Result Type</dt>
+   <dd>
+    <pre class="cddl">
+     EmptyResult
+    </pre>
+   </dd>
+</dl>
+
+<div algorithm="remote end steps for browsingContext.setUserAgent">
+
+The [=remote end steps=] with |command parameters| are:
+
+1. Let |context id| be the value of the <code>context</code> field of |command
+   parameters|.
+
+1. Let |context| be the result of [=trying=] to [=get a browsing context=] with
+   |context id|.
+
+1. Let |user agent| be the value of the <code>userAgent</code> field of |command
+   parameters| if present, or null otherwise.
+
+1. If |user agent| is not null:
+
+   1. Set |context|'s [=user agent=] to |user agent|.
```

Comment:
I am not sure if that would give use desired side effects? I am not sure that is really per-context because it is listed under "System state and capabilities". Also, the default User Agent value is used in https://fetch.spec.whatwg.org/#default-user-agent-value directly. I am not sure if we can redefine it from the BiDi spec or if we need some hooks into other spec. Perhaps we can use https://w3c.github.io/webdriver-bidi/#patches for that?

Reply:
I made some experiments and the user agent seems to be individual per context.

For example, given a page with an iframe, changing the userAgent in the iframe does not change it in the top-level context. And vice-versa.

Reply:
Hm, what do you mean by changing? CDP implementation is per-context but I don't think this spec change would specify that.

Reply:
Changing the user agent = calling https://chromedevtools.github.io/devtools-protocol/tot/Emulation/#method-setUserAgentOverride

Reply:
I'm not quite sure what is the AI here. Should I just change the prose to mention `fetch`?

Or should we restrict it to top-level contexts only?

Reply:
@jgraham what do you think of the approach proposed by @OrKoN? Should I look into it?

Reply:
So FWIW it looks like gecko _does_ restrict to the top level context. Obviously that could be changed if there's a use case for it.

I think the patch that would make sense would be to have a UA override defined on a (top-level) navigable. Then when asked to get the user agent in a specific Window you get the corresponding navigable and then (depending on the semantics we want) either go directly to it's top-level traversable and check if that has an override set, or maybe walk the parent tree looking for something with an override set. If there's no override you fall back to the default.

I think that works for requests that are kicked off from navigations or scripts running in a window. I don't know what implementations currently do for scripts running in workers, especially service workers or shared workers which could be associated with > 1 (top-level) navigable.

Reply:
I would find it hard to differentiate which override applies to all browsing contexts of a tab or just a specific one when we have more emulation. Maybe we should consider to only set it per top-level browsing context and that the setting bubbles down into frames as well? Here I don't see a reason why a frame should have a different user agent. Do you have an example for such a requirement?